src/builder.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{
+    config::Config,
+    value::Value,
+    source::Source,
+    error,
+    error::ConfigError,
+    path::Expression
+};
+use std::collections::HashMap;
+
+/**
+A configuration builder
+
+It registers ordered sources of configuration to later build consistent [`Config`] from them. 
+Configuration sources it defines are defaults, [`Source`]s and overrides.
+
+Defaults are alaways loaded first and can be overwritten by any of two other sources.
+Overrides are always loaded last, thus cannot be overridden. 
+Both can be only set explicitly key by key in code 
+using [`set_default`](Self::set_default) or [`set_override`](Self::set_override).
+
+An intermediate category, [`Source`], set groups of keys at once implicitly using data coming from external sources
+like files, environment variables or others that one implements. Defining a [`Source`] is as simple as implementing
+a trait for a struct.
+
+Adding sources, setting defaults and overrides does not invoke any I/O nor builds a config.
+It happens on demand when [`build`](Self::build) is called. 
+Therefore all errors, related to any of the [`Source`] will only show up then.
+```rust
+# use config::*;
+
+let mut builder = ConfigBuilder::default();
+
+builder.set_default("default", "1").expect("Key is valid");
+builder.add_source(File::new("config/settings", FileFormat::Json));
+builder.set_override("override", "1").expect("Key is valid");
+
+match builder.build() {
+    Ok(config) => {
+        // use your config
+    },
+    Err(e) => {
+        // something went wrong
+    }
+}
+```
```

Comment:
Two things: I like that the builder takes `&mut self`, this way it can be chained and called like you did in the code example above. I'd like to ask, though, whether you could add another example where you chain the calls.

Also, I think it would be a good idea to provide a `Config::builder()` function, as layed out in #180 for example. Good for discoverability and not too much overhead on our side.

Reply:
Done :)

Reply:
:+1: awesome. But you need to `build()` it, too! :laughing: